tensorflow/contrib/seastar/seastar_engine.cc

```diff
@@ -0,0 +1,170 @@
+namespace {
+  const static int DEFAULT_CORE_NUM = 4;
+  const static int kWaitTimeInUs = 50000;
+  const static int kRetryCount = 100;
+
+  // sesatar can't use localhost directly
+  string LocalhostToIp(const string& ip) {
+    const auto& vec = str_util::Split(ip, ':');
+    if (vec[0] == "localhost") {
+      return strings::StrCat(string("127.0.0.1:"), vec[1]);
+    } else {
+      return ip;
+    }
+  }
+
+  size_t GetCoreNumber(size_t total_connections) {
+    int64 max_core_number = DEFAULT_CORE_NUM;
+    Status s = ReadInt64FromEnvVar("SEASTAR_CORE_NUMBER", DEFAULT_CORE_NUM,
+                                   &max_core_number);
+
+    return std::min(total_connections, (size_t)max_core_number);
+  }
+
+  // force to avoid cleanup static variables and global variables in seastar
+  // engine which would trigger core
+  void SeastarExit(int status, void* arg) {
+    _exit(status);
+  }
+
+  void WaitForReady(const std::atomic_bool& a, int retry = kRetryCount) {
+    while (!a.load() && retry > 0) {
+      retry--;
+      usleep(kWaitTimeInUs);
+    }
+
+    if (!a.load()) {
+      LOG(FATAL) << "Seastar initialization failure!";
+    }
+  }
+}
+
+SeastarEngine::SeastarEngine(uint16_t server_number, uint16_t local,
+                             SeastarWorkerService* worker_service)
+  : local_(local), core_id_(0), is_server_ready_(false) {
+    ::on_exit(SeastarExit, (void*)nullptr);
+    tag_factory_ = new SeastarTagFactory(worker_service);
+    core_number_ = GetCoreNumber(server_number);
+    client_ = new SeastarClient();
+    thread_ = std::thread(&SeastarEngine::AsyncStartServer, this);
+}
+
+SeastarEngine::~SeastarEngine() {
+  /*
+    Don't need to cleanup the variables, only one SeastarEngine
+    These could be cleanup by process finished.
+  */
+}
+
+seastar::channel* SeastarEngine::GetChannel(const std::string& ip) {
+  WaitForReady(is_server_ready_);
+  seastar::channel* ch = AsyncConnect(ip);
+  WaitForReady(ch->is_init());
+  return ch;
+}
+
+seastar::channel* SeastarEngine::AsyncConnect(const std::string& ip) {
+  size_t core_id = core_id_ ++ % core_number_;
+  string s = LocalhostToIp(ip);
+  auto ch = new seastar::channel(s);
+  alien::submit_to(core_id, [core_id, s, ch, this] {
+    VLOG(2) << "client start connect core:" << core_id
+            << ", connect server:" << s;
+    client_->Connect(seastar::ipv4_addr{s}, s, ch, tag_factory_);
+    return seastar::make_ready_future();
+  });
+  return ch; 
+}
+
+/*
+  Simple solution of get cpuset, which could be more elegant
+  when k8s could provid the information.
+*/
+void SeastarEngine::GetCpuset(char** av) {
+  if (cpuset_.empty()) {
+    CpusetAllocator cpuset_alloc;
+    cpuset_ = cpuset_alloc.GetCpuset(core_number_);
+  }
+  if (cpuset_.empty()) {
+    LOG(FATAL) << "Internal error when launch grpc+seastar protocol,"
+               << "Please try other protocal";
+  }
+
+  *av = new char[cpuset_.size() + 1];
+  memcpy(*av, cpuset_.c_str(), cpuset_.size());
```

Comment:
copy '\0' by setting n = cpuset_.size() + 1 for memcopy ?

Reply:
thx @lxl910915, fixed.

Reply:
@liutongxuan Which version of hwloc you use ?

Reply:
> @liutongxuan Which version of hwloc do you use ?

In the pr, i disabled hwloc in seastar in thirdparty/seastar.BUILD. I'll open it later until the hwloc version is aligned between seastar & TensorFlow.

Reply:
@liutongxuan  We pull the latest code, an error occurs on worker, but we have no idea about it. Can you provide any suggestion to resolve it.
```
terminate called after throwing an instance of 'boost::exception_detail::clone_impl<boost::exception_detail::error_info_injector<boost::system::system_error> >'
  what():  Invalid argument
Aborted (core dumped)
```

Reply:
For error: "Seastar initialization failure!", seems can't connect some IPs. All IPs in cluster spec should exist in .endpoint_map.
For error: crash, can you show me the call stack of core dump?

Reply:
> For error: "Seastar initialization failure!", seems can't connect some IPs. All IPs in cluster spec should exist in .endpoint_map.
> For error: crash, can you show me the call stack of core dump?

We just run 1 ps and 1 worker on same machine, so we use localhost or 127.0.0.1 in cluster spec. And no core dump file was produced. Can you provide a building enviroment, such as a docker image? thx !

Reply:
You have changes based on the pr? what about use the pr without any changes, error or not?
Let me know your changes details?

From your error msg, generated `Aborted (core dumped)`. Please check the core dump directory.

Reply:
> You have changes based on the pr? what about use the pr without any changes, error or not?
> Let me know your changes details?
> 
> From your error msg, generated `Aborted (core dumped)`. Please check the core dump directory.

Using this pr, the bt in core dump is 
```
(gdb) bt
#0  0x00007f50b27da428 in __GI_raise (sig=sig@entry=6)
    at ../sysdeps/unix/sysv/linux/raise.c:54
#1  0x00007f50b27dc02a in __GI_abort () at abort.c:89
#2  0x00007f509c735249 in tensorflow::internal::LogMessageFatal::~LogMessageFatal (this=0x7f4f377fcdc0, __in_chrg=<optimized out>, __vtt_parm=<optimized out>)
    at tensorflow/core/platform/default/logging.cc:255
#3  0x00007f509477213b in tensorflow::(anonymous namespace)::WaitForReady (
    a=..., retry=0) at tensorflow/contrib/seastar/seastar_engine.cc:51
#4  0x00007f5094772508 in tensorflow::SeastarEngine::GetChannel (
    this=0x55f3e03c5510, ip="127.0.0.1:1112")
    at tensorflow/contrib/seastar/seastar_engine.cc:77
#5  0x00007f5094766036 in tensorflow::(anonymous namespace)::SparseSeastarChannelCache::FindChannelOnce (this=0x55f3e03c5ad0, 
    target="/job:ps/replica:0/task:0")
    at tensorflow/contrib/seastar/seastar_channel_cache.cc:195
#6  0x00007f5094764fa0 in tensorflow::(anonymous namespace)::CachingSeastarChannelCache::FindWorkerChannel (this=0x55f3e03c5ad0, 
    target="/job:ps/replica:0/task:0")
    at tensorflow/contrib/seastar/seastar_channel_cache.cc:63
#7  0x00007f50947656f9 in tensorflow::(anonymous namespace)::MultiSeastarChannelCache::FindChannelOnce (this=0x55f3e03c5d00, target="/job:ps/replica:0/task:0")
    at tensorflow/contrib/seastar/seastar_channel_cache.cc:123
#8  0x00007f5094764fa0 in tensorflow::(anonymous namespace)::CachingSeastarChann---Type <return> to continue, or q <return> to quit---
elCache::FindWorkerChannel (this=0x55f3e03c5d00, 
    target="/job:ps/replica:0/task:0")
    at tensorflow/contrib/seastar/seastar_channel_cache.cc:63
#9  0x00007f5094763105 in tensorflow::(anonymous namespace)::SeastarWorkerCache::CreateWorker (this=0x55f3e03c5e50, target="/job:ps/replica:0/task:0")
    at tensorflow/contrib/seastar/seastar_worker_cache.cc:38
#10 0x00007f5094b75fca in tensorflow::NewRemoteDevices(tensorflow::Env*, tensorflow::WorkerCacheInterface*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::function<void (tensorflow::Status const&, std::vector<tensorflow::Device*, std::allocator<tensorflow::Device*> >*)>) (
    env=0x55f3ddfddb60, worker_cache=0x55f3e03c5e50, 
    worker_name="/job:ps/replica:0/task:0", done=...)
    at tensorflow/core/distributed_runtime/remote_device.cc:59
#11 0x00007f5094b6c1ce in tensorflow::DeviceFinder::Start (this=0x7f4f377fd400)
    at tensorflow/core/distributed_runtime/master.cc:250
#12 0x00007f5094b6ab4f in tensorflow::DeviceFinder::GetRemoteDevices (
    device_filters=..., env=0x55f3e00cb8d8, worker_cache=0x55f3e03c5e50, 
    out_remote=0x7f4f200008c0)
    at tensorflow/core/distributed_runtime/master.cc:140
#13 0x00007f5094b6363d in tensorflow::Master::<lambda()>::operator()(void) const (__closure=0x55f3e6df5150)
    at tensorflow/core/distributed_runtime/master.cc:444
#14 0x00007f5094b67b3b in std::_Function_handler<void(), tensorflow::Master::Cre---Type <return> to continue, or q <return> to quit---
ateSession(const tensorflow::CreateSessionRequest*, tensorflow::CreateSessionResponse*, tensorflow::Master::MyClosure)::<lambda()> >::_M_invoke(const std::_Any_data &) (__functor=...) at /usr/include/c++/5/functional:1871
#15 0x00007f508a631078 in std::function<void ()>::operator()() const (
    this=0x55f3e6df5658) at /usr/include/c++/5/functional:2267
#16 0x00007f508ac548d0 in std::_Bind_simple<std::function<void ()> ()>::_M_invoke<>(std::_Index_tuple<>) (this=0x55f3e6df5658)
    at /usr/include/c++/5/functional:1531
#17 0x00007f508ac5486d in std::_Bind_simple<std::function<void ()> ()>::operator()() (this=0x55f3e6df5658) at /usr/include/c++/5/functional:1520
#18 0x00007f508ac5480c in std::thread::_Impl<std::_Bind_simple<std::function<void ()> ()> >::_M_run() (this=0x55f3e6df5640) at /usr/include/c++/5/thread:115
#19 0x00007f5087accc80 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#20 0x00007f50b2b766ba in start_thread (arg=0x7f4f377fe700)
    at pthread_create.c:333
#21 0x00007f50b28ac41d in clone ()
    at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109
```
And .endpoint_map is 
```
127.0.0.1:1111=127.0.0.1:1112
127.0.0.1:2222=127.0.0.1:2223
```

Reply:
Seastar client try to connect 127.0.0.1:1112 failed. It seems the peer's service haven't launched, can you show me the peer's log (which listen 127.0.0.1:1112)? 
Please use " TF_CPP_MIN_VLOG_LEVEL=2 " before you launch command to show more logs.

Reply:
> Seastar client try to connect 127.0.0.1:1112 failed. It seems the peer's service haven't launched, can you show me the peer's log (which listen 127.0.0.1:1112)?
> Please use " TF_CPP_MIN_VLOG_LEVEL=2 " before you launch command to show more logs.

On ps, seastar start successfully with log '[seastar_engine.cc:164] Seastar server started successfully, listen port: 1112' .

On worker, it hang at connecting to ps with log '[seastar_engine.cc:85] client start connect core:0, connect server:127.0.0.1:1112' .